File: ui/src/lib/api/sources.ts
"use client";

/**
 * Sources API: cron-scheduled fetches that feed a pipeline.
 *
 * Mirrors `crates/gateway/src/sources.rs` (`SourceView`, `CreateSource`,
 * `UpdateSource`, `RunRecord`), `crates/source/src/model.rs` (`Location`,
 * `FetchAuth`, `RunOutcome`) and `crates/gateway/src/handlers/sources.rs`.
 *
 * Like connections, every route answers `501 not_configured` until the
 * deployment sets `SOURCE_SECRET_KEY`.
 */
import {
  useMutation,
  useQuery,
  useQueryClient,
  type UseMutationResult,
  type UseQueryResult,
} from "@tanstack/react-query";
import { toast } from "sonner";

import { isFormError, request } from "./client";
import { errorMessage } from "./hooks";
import { jobQueryKeys } from "./jobs";
import type { PipelineDefinition } from "./types";

// ---------------------------------------------------------------------------
// Wire types
// ---------------------------------------------------------------------------

/** `Location` — a tagged union; `url` is the only variant in v1. */
export interface UrlLocation {
  kind: "url";
  /** May contain `{{ date:FMT }}`, `{{ date-1d:FMT }}`, `{{ timestamp }}`. */
  url: string;
  /** `GET` when absent. */
  method?: string;
  /** Non-secret headers. Secrets belong in `auth`. */
  headers?: Record<string, string>;
}

/** Where a source fetches from. */
export type SourceLocation = UrlLocation;

/** `FetchAuth` as it is written: values in clear, sealed by the gateway. */
export type FetchAuth =
  | { kind: "bearer"; token: string }
  | { kind: "basic"; username: string; password: string }
  | { kind: "headers"; headers: Record<string, string> };

/** Every credential kind, in the order the form offers them. */
export const FETCH_AUTH_KINDS = ["bearer", "basic", "headers"] as const;
/** `FetchAuth["kind"]`. */
export type FetchAuthKind = (typeof FETCH_AUTH_KINDS)[number];

/**
 * `FetchAuth` as it is read back: every secret value is `"****"`, the basic
 * username and the header names are kept. `unknown` is what the gateway answers
 * when it can no longer open the sealed blob.
 */
export type RedactedFetchAuth =
  | { kind: "bearer"; token: string }
  | { kind: "basic"; username: string; password: string }
  | { kind: "headers"; headers: Record<string, string> }
  | { kind: "unknown"; sealed?: boolean };

/** `RunOutcome`. */
export type RunOutcome = "ingested" | "unchanged" | "failed";

/** `SourceView`. Timestamps are RFC 3339. */
export interface SourceView {
  uid: string;
  name: string;
  description?: string;
  project_id?: string;
  /** Pipeline uid fed on every tick. */
  pipeline: string;
  location: SourceLocation;
  cron: string;
  /** IANA timezone for the cron and the URL template. */
  timezone: string;
  paused: boolean;
  /** Index override. */
  index?: string;
  /** Redacted credential, when one is set. */
  auth?: RedactedFetchAuth;
  /** Set when the pipeline was deleted; archived sources never fire. */
  archived_at?: string;
  /** Next scheduled run — single-source reads only. */
  next_run_at?: string;
  last_run_at?: string;
  last_status?: RunOutcome;
  last_error?: string;
}

/** `POST /sources` body (`deny_unknown_fields`: send nothing else). */
export interface CreateSourceBody {
  uid: string;
  name?: string;
  description?: string;
  pipeline: string;
  location: SourceLocation;
  cron: string;
  timezone?: string;
  index?: string;
  auth?: FetchAuth;
}

/**
 * `PATCH /sources/{uid}` body. `auth` absent keeps the credential, `null`
 * clears it, an object replaces it.
 */
export interface UpdateSourceBody {
  name?: string;
  description?: string;
  pipeline?: string;
  location?: SourceLocation;
  cron?: string;
  timezone?: string;
  index?: string;
  auth?: FetchAuth | null;
}

/** One row of `GET /sources/{uid}/runs`, newest first. */
export interface RunRecord {
  run_id: string;
  source_id: string;
  started_at: string;
  finished_at?: string;
  outcome: RunOutcome;
  /** Items fetched (documents or files handed to the pipeline). */
  items: number;
  /** Jobs this run started; empty for `unchanged` and `failed`. */
  job_ids: string[];
  error?: string;
}

/** 202 body of `POST /sources/{uid}/run`. */
export interface TriggerRunResponse {
  status: string;
}

// ---------------------------------------------------------------------------
// Pure helpers (unit-tested in ./sources.test.ts)
// ---------------------------------------------------------------------------

/** The plugin whose `config.connection` pins a pipeline's destination. */
export const INDEXER_PLUGIN = "meili_indexer";

/**
 * Mirrors `PipelineDefinition::pins_destination()` on the Rust side: the
 * pipeline has at least one `meili_indexer` step and every one of them names a
 * non-blank `connection`. A scheduled run has no request to supply a
 * Meilisearch, so this is what the gateway requires of a source's pipeline.
 */
export function pipelinePinsConnection(pipeline: PipelineDefinition): boolean {
  const indexers = pipeline.steps.filter((step) => step.plugin === INDEXER_PLUGIN);
  if (indexers.length === 0) return false;
  return indexers.every((step) => {
    const connection = step.config?.connection;
    return typeof connection === "string" && connection.trim().length > 0;
  });
}

/** Whether "run now" and pause/unpause make sense for this source. */
export function isArchived(source: Pick<SourceView, "archived_at">): boolean {
  return Boolean(source.archived_at);
}

// ---------------------------------------------------------------------------
// API calls
// ---------------------------------------------------------------------------

function sourcePath(uid: string): string {
  return `/sources/${encodeURIComponent(uid)}`;
}

/**
 * Query string for `GET /sources`, leading `?` included (empty when there is
 * nothing to send). The gateway omits archived sources unless asked.
 */
export function sourceListSearch(options: { includeArchived: boolean }): string {
  return options.includeArchived ? "?include_archived=true" : "";
}

/** `GET /sources[?include_archived=true]`. */
export function listSources(
  options: { includeArchived: boolean },
  signal?: AbortSignal,
): Promise<SourceView[]> {
  return request<SourceView[]>(`/sources${sourceListSearch(options)}`, { signal });
}

/** `GET /sources/{uid}` — includes `next_run_at`. */
export function getSource(uid: string, signal?: AbortSignal): Promise<SourceView> {
  return request<SourceView>(sourcePath(uid), { signal });
}

/** `POST /sources`. */
export function createSource(body: CreateSourceBody): Promise<SourceView> {
  return request<SourceView>("/sources", { method: "POST", json: body });
}

/** `PATCH /sources/{uid}`. */
export function updateSource(uid: string, body: UpdateSourceBody): Promise<SourceView> {
  return request<SourceView>(sourcePath(uid), { method: "PATCH", json: body });
}

/** `DELETE /sources/{uid}` — removes the schedule, then the row. */
export function deleteSource(uid: string): Promise<void> {
  return request<void>(sourcePath(uid), { method: "DELETE" });
}

/** `POST /sources/{uid}/pause` or `/unpause`. */
export function setSourcePaused(uid: string, paused: boolean): Promise<void> {
  return request<void>(`${sourcePath(uid)}/${paused ? "pause" : "unpause"}`, { method: "POST" });
}

/** `POST /sources/{uid}/run` — 202; 422 on an archived source. */
export function triggerSourceRun(uid: string): Promise<TriggerRunResponse> {
  return request<TriggerRunResponse>(`${sourcePath(uid)}/run`, { method: "POST" });
}

/** Default page of run history. */
export const SOURCE_RUNS_LIMIT = 20;

/** `GET /sources/{uid}/runs?limit=` — newest first. */
export function listSourceRuns(
  uid: string,
  limit: number = SOURCE_RUNS_LIMIT,
  signal?: AbortSignal,
): Promise<RunRecord[]> {
  const query = new URLSearchParams({ limit: String(limit) });
  return request<RunRecord[]>(`${sourcePath(uid)}/runs?${query.toString()}`, { signal });
}

// ---------------------------------------------------------------------------
// Query bindings
// ---------------------------------------------------------------------------

/** Cache keys for everything source-shaped. */
export const sourceQueryKeys = {
  all: ["sources"] as const,
  list: (includeArchived: boolean) => ["sources", "list", { includeArchived }] as const,
  detail: (uid: string) => ["sources", "detail", uid] as const,
  runs: (uid: string) => ["sources", "runs", uid] as const,
};

/** Every source visible to the caller, archived ones only when asked. */
export function useSources(
  includeArchived: boolean = true,
): UseQueryResult<SourceView[], Error> {
  return useQuery({
    queryKey: sourceQueryKeys.list(includeArchived),
    queryFn: ({ signal }) => listSources({ includeArchived }, signal),
    placeholderData: (previous) => previous,
  });
}

/** One source, with `next_run_at`. Disabled while `uid` is empty. */
export function useSource(uid: string | undefined): UseQueryResult<SourceView, Error> {
  return useQuery({
    queryKey: sourceQueryKeys.detail(uid ?? ""),
    queryFn: ({ signal }) => getSource(uid as string, signal),
    enabled: Boolean(uid),
  });
}

/** Recent runs of one source. Disabled while `uid` is empty. */
export function useSourceRuns(uid: string | undefined): UseQueryResult<RunRecord[], Error> {
  return useQuery({
    queryKey: sourceQueryKeys.runs(uid ?? ""),
    queryFn: ({ signal }) => listSourceRuns(uid as string, SOURCE_RUNS_LIMIT, signal),
    enabled: Boolean(uid),
  });
}

/**
 * Refresh everything a change to one source can move: the list, its detail and
 * its runs. `sourceQueryKeys.all` is a prefix of all three.
 */
async function invalidateSources(
  queryClient: ReturnType<typeof useQueryClient>,
): Promise<void> {
  await queryClient.invalidateQueries({ queryKey: sourceQueryKeys.all });
}

/**
 * Create a source. A 422 (bad cron, unpinned pipeline, host outside the fetch
 * policy) is left to the form; anything else is a toast.
 */
export function useCreateSource(): UseMutationResult<SourceView, Error, CreateSourceBody> {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (body: CreateSourceBody) => createSource(body),
    onSuccess: async (saved) => {
      toast.success(`Source "${saved.uid}" created`, { description: saved.cron });
      await invalidateSources(queryClient);
    },
    onError: (error) => {
      if (isFormError(error)) return;
      toast.error("Could not create the source", { description: errorMessage(error) });
    },
  });
}

/** Update a source. Same failure contract as {@link useCreateSource}. */
export function useUpdateSource(): UseMutationResult<
  SourceView,
  Error,
  { uid: string; body: UpdateSourceBody }
> {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ uid, body }) => updateSource(uid, body),
    onSuccess: async (saved) => {
      toast.success(`Source "${saved.uid}" saved`);
      await invalidateSources(queryClient);
    },
    onError: (error) => {
      if (isFormError(error)) return;
      toast.error("Could not save the source", { description: errorMessage(error) });
    },
  });
}

/** Delete a source. Drops its detail and runs, refreshes the list. */
export function useDeleteSource(): UseMutationResult<void, Error, string> {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (uid: string) => deleteSource(uid),
    onSuccess: async (_result, uid) => {
      toast.success(`Source "${uid}" deleted`);
      queryClient.removeQueries({ queryKey: sourceQueryKeys.detail(uid) });
      queryClient.removeQueries({ queryKey: sourceQueryKeys.runs(uid) });
      await invalidateSources(queryClient);
    },
    onError: (error) => {
      toast.error("Could not delete the source", { description: errorMessage(error) });
    },
  });
}

/** Pause or resume a source's schedule. */
export function useSetSourcePaused(): UseMutationResult<
  void,
  Error,
  { uid: string; paused: boolean }
> {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ uid, paused }) => setSourcePaused(uid, paused),
    onSuccess: async (_result, { uid, paused }) => {
      toast.success(paused ? `Source "${uid}" paused` : `Source "${uid}" resumed`);
      await invalidateSources(queryClient);
    },
    onError: (error, { paused }) => {
      toast.error(paused ? "Could not pause the source" : "Could not resume the source", {
        description: errorMessage(error),
      });
    },
  });
}

/**
 * Run a source now. The run is asynchronous (202): the source, its runs and the
 * jobs list are invalidated so the screens pick the run up as soon as it lands.
 */
export function useTriggerSourceRun(): UseMutationResult<TriggerRunResponse, Error, string> {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (uid: string) => triggerSourceRun(uid),
    onSuccess: async (_response, uid) => {
      toast.success(`Run of "${uid}" triggered`, {
        description: "It appears in the run history once the fetch finishes.",
      });
      await Promise.all([
        invalidateSources(queryClient),
        queryClient.invalidateQueries({ queryKey: jobQueryKeys.all }),
      ]);
    },
    onError: (error) => {
      toast.error("Could not trigger a run", { description: errorMessage(error) });
    },
  });
}

File: crates/gateway/src/sources.rs
//! Scheduled sources, gateway side: the control-plane calls, save-time validation, and
//! the redacted public shape. The routes live in [`crate::handlers::sources`].
//!
//! The gateway seals a source's fetch credential before it reaches the control plane
//! and never returns it; a source's *destination* is not stored here at all — it is the
//! Meilisearch connection its pipeline's `meili_indexer` step names (spec Decision 6).

use chrono::{DateTime, Utc};
use meili_ingest_plugin_sdk::PipelineDefinition;
use meili_ingest_source::{
    FetchAuth, HostPolicy, IncrementalState, Location, RunOutcome, SecretKey, SourceDefinition,
    SourceError, open_json, redact, render,
};
use serde::{Deserialize, Serialize};
use url::Url;
use uuid::Uuid;

use crate::error::GatewayError;
use crate::state::ControlPlaneClient;

// ---------------------------------------------------------------------------
// Control-plane wire types (mirror `meili-ingest-control-plane::sources`).
// ---------------------------------------------------------------------------

/// A source as the control plane stores it; the credential is still sealed.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct SourceRecord {
    /// The public shape.
    #[serde(flatten)]
    pub definition: SourceDefinition,
    /// Sealed fetch credential.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub fetch_auth: Option<Vec<u8>>,
    /// What the previous run learned.
    #[serde(default)]
    pub state: IncrementalState,
    /// When the last run started.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub last_run_at: Option<DateTime<Utc>>,
    /// Outcome of the last run.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub last_status: Option<String>,
    /// Error of the last run.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub last_error: Option<String>,
}

/// One run of a source.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RunRecord {
    /// Run id.
    pub run_id: Uuid,
    /// Source this run belongs to.
    pub source_id: Uuid,
    /// When the run started.
    pub started_at: DateTime<Utc>,
    /// When it finished.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub finished_at: Option<DateTime<Utc>>,
    /// Terminal outcome.
    pub outcome: RunOutcome,
    /// Source items ingested.
    #[serde(default)]
    pub items: i32,
    /// Jobs started.
    #[serde(default)]
    pub job_ids: Vec<Uuid>,
    /// Failure message.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

/// Body of `POST /internal/sources`.
#[derive(Debug, Clone, Serialize)]
pub struct NewSource {
    /// Surrogate id, which also names the schedule.
    pub id: Uuid,
    /// Handle.
    pub uid: String,
    /// Display name.
    pub name: String,
    /// Description.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    /// Tenant scope.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub project_id: Option<String>,
    /// Pipeline fed.
    pub pipeline_uid: String,
    /// Where to fetch.
    pub location: Location,
    /// Cron.
    pub cron: String,
    /// Timezone.
    pub timezone: String,
    /// Index override.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub index_name: Option<String>,
    /// Sealed credential.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub fetch_auth: Option<Vec<u8>>,
    /// Temporal schedule id.
    pub schedule_id: String,
}

/// Body of `PATCH /internal/sources/{uid}`. `fetch_auth: Some(None)` serializes as
/// `null`, which the control plane reads as "clear".
#[derive(Debug, Clone, Default, Serialize)]
pub struct SourcePatch {
    /// New name.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    /// New description.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    /// New pipeline.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pipeline_uid: Option<String>,
    /// New location.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub location: Option<Location>,
    /// New cron.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cron: Option<String>,
    /// New timezone.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub timezone: Option<String>,
    /// New index override.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub index_name: Option<String>,
    /// Replace (`Some(Some)`), clear (`Some(None)`) or keep (`None`) the credential.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub fetch_auth: Option<Option<Vec<u8>>>,
    /// New paused flag.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub paused: Option<bool>,
}

// ---------------------------------------------------------------------------
// Public API shapes.
// ---------------------------------------------------------------------------

fn utc() -> String {
    "UTC".to_string()
}

/// `POST /sources` body.
#[derive(Debug, Clone, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct CreateSource {
    /// Handle, unique per project.
    pub uid: String,
    /// Display name; defaults to the uid.
    #[serde(default)]
    pub name: Option<String>,
    /// Description.
    #[serde(default)]
    pub description: Option<String>,
    /// Pipeline to run on every tick. Its `meili_indexer` must name a connection.
    pub pipeline: String,
    /// Where to fetch from.
    pub location: Location,
    /// Cron expression, validated by Temporal.
    pub cron: String,
    /// IANA timezone for the cron and for URL templates.
    #[serde(default = "utc")]
    pub timezone: String,
    /// Index override.
    #[serde(default)]
    pub index: Option<String>,
    /// Credential for the fetch. Sealed before it leaves the gateway.
    #[serde(default)]
    pub auth: Option<FetchAuth>,
}

/// `PATCH /sources/{uid}` body. `auth` omitted keeps it, `null` clears it.
#[derive(Debug, Clone, Default, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct UpdateSource {
    /// New name.
    #[serde(default)]
    pub name: Option<String>,
    /// New description.
    #[serde(default)]
    pub description: Option<String>,
    /// New pipeline.
    #[serde(default)]
    pub pipeline: Option<String>,
    /// New location.
    #[serde(default)]
    pub location: Option<Location>,
    /// New cron.
    #[serde(default)]
    pub cron: Option<String>,
    /// New timezone.
    #[serde(default)]
    pub timezone: Option<String>,
    /// New index override.
    #[serde(default)]
    pub index: Option<String>,
    /// Replace, clear (`null`) or keep (absent) the credential.
    #[serde(default, deserialize_with = "double_option")]
    pub auth: Option<Option<FetchAuth>>,
}

fn double_option<'de, D, T>(d: D) -> Result<Option<Option<T>>, D::Error>
where
    D: serde::Deserializer<'de>,
    T: Deserialize<'de>,
{
    Option::<T>::deserialize(d).map(Some)
}

/// A source as the API returns it. The credential is shown only as its kind and header
/// names, never a value.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SourceView {
    /// Handle.
    pub uid: String,
    /// Display name.
    pub name: String,
    /// Description.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    /// Tenant scope.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub project_id: Option<String>,
    /// Pipeline fed.
    pub pipeline: String,
    /// Where it fetches from.
    pub location: Location,
    /// Cron.
    pub cron: String,
    /// Timezone.
    pub timezone: String,
    /// Whether the schedule is paused.
    pub paused: bool,
    /// Index override.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub index: Option<String>,
    /// Redacted credential, when one is set.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub auth: Option<serde_json::Value>,
    /// Set when the source's pipeline was deleted; archived sources never fire.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub archived_at: Option<DateTime<Utc>>,
    /// Next scheduled run, on single-source reads.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub next_run_at: Option<DateTime<Utc>>,
    /// When the last run started.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub last_run_at: Option<DateTime<Utc>>,
    /// Outcome of the last run.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub last_status: Option<String>,
    /// Error of the last run.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub last_error: Option<String>,
}

impl SourceView {
    /// The public view of a record. `key` opens the credential only to learn its kind.
    pub fn from_record(
        r: SourceRecord,
        key: &SecretKey,
        next_run_at: Option<DateTime<Utc>>,
    ) -> Self {
        let auth = r.fetch_auth.as_deref().map(|sealed| {
            match open_json::<FetchAuth>(key, sealed) {
                Ok(auth) => redact(&auth),
                // Still say a credential exists without pretending to know its kind.
                Err(_) => serde_json::json!({ "kind": "unknown", "sealed": true }),
            }
        });
        let d = r.definition;
        Self {
            uid: d.uid,
            name: d.name,
            description: d.description,
            project_id: d.project_id,
            pipeline: d.pipeline_uid,
            location: d.location,
            cron: d.cron,
            timezone: d.timezone,
            paused: d.paused,
            index: d.index_name,
            auth,
            archived_at: d.archived_at,
            next_run_at,
            last_run_at: r.last_run_at,
            last_status: r.last_status,
            last_error: r.last_error,
        }
    }
}

// ---------------------------------------------------------------------------
// Validation.
// ---------------------------------------------------------------------------

/// Check a location can be fetched: its template renders (known tokens, known
/// timezone), the result is a URL, and the host passes `SOURCE_FETCH_HOSTS`. The fetch
/// re-checks the host at every run and every redirect; this is the early, friendly
/// failure.
pub async fn validate_location(
    location: &Location,
    timezone: &str,
    policy: &HostPolicy,
) -> Result<(), GatewayError> {
    let Location::Url { url, .. } = location;
    let rendered = render(url, Utc::now(), timezone).map_err(unprocessable)?;
    let parsed = Url::parse(&rendered).map_err(|e| {
        GatewayError::Unprocessable(format!("location {rendered:?} is not a URL: {e}"))
    })?;
    policy.check(&parsed).await.map_err(unprocessable)
}

fn unprocessable(e: SourceError) -> GatewayError {
    GatewayError::Unprocessable(e.to_string())
}

/// A scheduled run has no request, so the pipeline must supply its own destination.
pub fn require_pinned(pipeline: &PipelineDefinition) -> Result<(), GatewayError> {
    if pipeline.pins_destination() {
        Ok(())
    } else {
        Err(GatewayError::Unprocessable(format!(
            "pipeline {:?} cannot run on a schedule: its meili_indexer step names no \
             Meilisearch `connection`, and a scheduled run has no request to supply a \
             destination",
            pipeline.uid
        )))
    }
}

// ---------------------------------------------------------------------------
// Control-plane calls.
// ---------------------------------------------------------------------------

fn reword(e: GatewayError) -> GatewayError {
    match e {
        GatewayError::Invalid(msg) => GatewayError::Unprocessable(msg),
        other => other,
    }
}

fn not_found_to_none<T>(r: Result<T, GatewayError>) -> Result<Option<T>, GatewayError> {
    match r {
        Ok(v) => Ok(Some(v)),
        Err(GatewayError::NotFound(_)) => Ok(None),
        Err(e) => Err(e),
    }
}

impl ControlPlaneClient {
    fn source_url(&self, prefix: &str, key: &str, suffix: &str) -> Result<Url, GatewayError> {
        let mut url = Url::parse(&self.url(&format!("/internal/{prefix}/")))
            .map_err(|e| GatewayError::Internal(format!("control plane url: {e}")))?;
        {
            let mut segments = url.path_segments_mut().map_err(|()| {
                GatewayError::Internal("control plane url cannot be a base".into())
            })?;
            segments.pop_if_empty().push(key);
            if !suffix.is_empty() {
                segments.push(suffix);
            }
        }
        Ok(url)
    }

    /// `GET /internal/sources`.
    pub async fn list_sources(
        &self,
        project_id: Option<&str>,
        include_archived: bool,
    ) -> Result<Vec<SourceRecord>, GatewayError> {
        let mut query = Self::project_query(project_id);
        if include_archived {
            query.push(("include_archived", "true".into()));
        }
        let req = self.http.get(self.url("/internal/sources")).query(&query);
        self.send_json(req, "list sources").await
    }

    /// `GET /internal/sources/{uid}`; `Ok(None)` when missing.
    pub async fn get_source(
        &self,
        uid: &str,
        project_id: Option<&str>,
    ) -> Result<Option<SourceRecord>, GatewayError> {
        let req = self
            .http
            .get(self.source_url("sources", uid, "")?)
            .query(&Self::project_query(project_id));
        not_found_to_none(self.send_json(req, "get source").await)
    }

    /// `POST /internal/sources`.
    pub async fn create_source(&self, new: &NewSource) -> Result<SourceRecord, GatewayError> {
        let req = self.http.post(self.url("/internal/sources")).json(new);
        self.send_json(req, "create source").await.map_err(reword)
    }

    /// `PATCH /internal/sources/{uid}`; `Ok(None)` when not in exactly this scope.
    pub async fn update_source(
        &self,
        uid: &str,
        project_id: Option<&str>,
        patch: &SourcePatch,
    ) -> Result<Option<SourceRecord>, GatewayError> {
        let req = self
            .http
            .patch(self.source_url("sources", uid, "")?)
            .query(&Self::project_query(project_id))
            .json(patch);
        not_found_to_none(self.send_json(req, "update source").await).map_err(reword)
    }

    /// `DELETE /internal/sources/{uid}`.
    pub async fn delete_source(
        &self,
        uid: &str,
        project_id: Option<&str>,
    ) -> Result<(), GatewayError> {
        let req = self
            .http
            .delete(self.source_url("sources", uid, "")?)
            .query(&Self::project_query(project_id));
        self.send_empty(req, "delete source").await
    }

    /// `PUT /internal/sources-by-id/{id}/paused`.
    pub async fn set_source_paused(&self, id: Uuid, paused: bool) -> Result<(), GatewayError> {
        let req = self
            .http
            .put(self.source_url("sources-by-id", &id.to_string(), "paused")?)
            .json(&serde_json::json!({ "paused": paused }));
        self.send_empty(req, "pause source").await
    }

    /// `GET /internal/sources-by-id/{id}/runs`.
    pub async fn list_source_runs(
        &self,
        id: Uuid,
        limit: i64,
    ) -> Result<Vec<RunRecord>, GatewayError> {
        let req = self
            .http
            .get(self.source_url("sources-by-id", &id.to_string(), "runs")?)
            .query(&[("limit", limit.to_string())]);
        self.send_json(req, "list source runs").await
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::BTreeMap;

    fn url(u: &str) -> Location {
        Location::Url {
            url: u.into(),
            method: None,
            headers: BTreeMap::new(),
        }
    }

    #[tokio::test]
    async fn an_unknown_template_token_is_rejected_on_save() {
        let err = validate_location(
            &url("https://x.example/{{ nope }}"),
            "UTC",
            &HostPolicy::Any,
        )
        .await
        .expect_err("unknown token");
        assert!(matches!(err, GatewayError::Unprocessable(_)), "{err:?}");
    }

    #[tokio::test]
    async fn an_unknown_timezone_is_rejected_on_save() {
        let err = validate_location(
            &url("https://x.example/a"),
            "Mars/Olympus",
            &HostPolicy::Any,
        )
        .await
        .expect_err("unknown tz");
        assert!(err.to_string().contains("timezone"), "{err}");
    }

    #[tokio::test]
    async fn the_fetch_policy_is_applied_on_save() {
        let err = validate_location(
            &url("http://files.internal/export.json"),
            "UTC",
            &HostPolicy::Public,
        )
        .await
        .expect_err("public-only rejects http");
        assert!(matches!(err, GatewayError::Unprocessable(_)), "{err:?}");
        validate_location(
            &url("http://files.internal/export_{{ date:%Y }}.json"),
            "UTC",
            &HostPolicy::parse("files.internal").expect("policy"),
        )
        .await
        .expect("an allowlisted host is accepted");
    }

    #[test]
    fn a_pipeline_without_a_connection_cannot_be_scheduled() {
        let mut p = PipelineDefinition {
            uid: "plain".into(),
            name: "plain".into(),
            description: None,
            version: 1,
            trigger: None,
            steps: vec![meili_ingest_plugin_sdk::StepDefinition::new(
                "index",
                "meili_indexer",
            )],
            builtin: false,
            project_id: None,
        };
        let err = require_pinned(&p).expect_err("unpinned");
        assert!(err.to_string().contains("connection"), "{err}");
        p.steps[0].config = serde_json::json!({ "connection": "prod" });
        require_pinned(&p).expect("pinned");
    }

    #[test]
    fn the_view_shows_the_credential_kind_but_never_its_value() {
        let key = SecretKey::from_bytes([2; 32]);
        let sealed = meili_ingest_source::seal_json(
            &key,
            &FetchAuth::Bearer {
                token: "s3cret".into(),
            },
        )
        .expect("seal");
        let record = SourceRecord {
            definition: SourceDefinition {
                id: Uuid::nil(),
                uid: "tmdb".into(),
                name: "TMDB".into(),
                description: None,
                project_id: None,
                pipeline_uid: "movies".into(),
                location: url("https://x.example/a"),
                cron: "0 3 * * *".into(),
                timezone: "UTC".into(),
                paused: false,
                index_name: None,
                schedule_id: "source-x".into(),
                archived_at: None,
            },
            fetch_auth: Some(sealed),
            state: IncrementalState::default(),
            last_run_at: None,
            last_status: None,
            last_error: None,
        };
        let view = SourceView::from_record(record, &key, None);
        let json = serde_json::to_string(&view).expect("serialize");
        assert_eq!(
            view.auth,
            Some(serde_json::json!({"kind": "bearer", "token": "****"}))
        );
        assert!(!json.contains("s3cret"), "{json}");
    }

    #[test]
    fn a_patch_body_distinguishes_clear_from_keep() {
        let clear: UpdateSource = serde_json::from_str(r#"{"auth": null}"#).expect("json");
        assert!(matches!(clear.auth, Some(None)), "null clears");
        let keep: UpdateSource = serde_json::from_str("{}").expect("json");
        assert!(keep.auth.is_none(), "absent keeps");
        let wire = SourcePatch {
            fetch_auth: Some(None),
            ..Default::default()
        };
        assert_eq!(
            serde_json::to_value(&wire).expect("json"),
            serde_json::json!({ "fetch_auth": null }),
            "the control plane receives an explicit null"
        );
    }
}
